### PR TITLE
[Tracing Redesign][2/N] Implement tracing unary Interceptors

### PR DESCRIPTION
**Pull request by kexiongliu123**
This PR introduces the unary interceptors in YARPC Go to improve the structure and functionality of tracing and middleware handling. The primary focus is on separating the unary interceptors to allow for specialized optimizations and cleaner handling of tracing operations.

Key Changes:

Introduced a dedicated UnaryInbound and UnaryOutbound middleware architecture, simplifying the tracing logic for unary requests.
Introduced inbound.go and outbound.go files to reflect the specialized unary middleware design, improving maintainability and reducing complexity.
Enhanced test coverage for the unary path, ensuring that all middleware logic is properly unit tested with specific focus on the flow of unary RPCs.
Updated helper methods for extracting and applying tracing tags to unary requests, ensuring clearer separation of concerns.

### CODE REVIEW

File: internal/tracinginterceptor/interceptor.go [hunk 1]
```diff
@@ -39,32 +44,76 @@ var (
...
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         tags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	err := h.Handle(ctx, req, resw)
+	return updateSpanWithError(span, err)
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:88]: For inbound/ handler, checking the error is probably not enough - for TChannel, there are failure cases where error is nil. And we probably need similar [writer wrapper implementation](https://github.com/yarpc/yarpc-go/blob/dev/internal/observability/middleware.go#L40) in observability middleware.

File: internal/tracinginterceptor/interceptor_test.go [hunk 2]
```diff
@@ -19,3 +19,143 @@
...
+	"context"
+	"testing"
+
+	"github.com/opentracing/opentracing-go/mocktracer"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/yarpc/api/transport"
+)
+
+// Define UnaryHandlerFunc to adapt a function into a UnaryHandler.
+type UnaryHandlerFunc func(ctx context.Context, req *transport.Request, resw transport.ResponseWriter) error
```
[the remark this answers is not in the document]
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor_test.go:33]: Using yarpctest.NewFakeTransport() may not be feasible, as the provided FakeTransport implementation from yarpctest lacks certain methods and overrides (e.g., NewInbound(), SetHandler()) needed to set up the required inbound or outbound call behaviors expected by the tracing interceptor. the interceptor requires interactions with the UnaryHandler or UnaryOutbound interface, and FakeTransport doesn't include built-in functionality to create appropriate instances of these interfaces.

File: transport/tchannel/tracing.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2024 Uber Technologies, Inc.
```
Comment (ChenX1993) [inline transport/tchannel/tracing.go:1]: I should have made it more clear - let's put it under `transport/tchannel/tracing/tracing.go`, otherwise it is still under the same Go package that will cause circular dependency later.

File: internal/tracinginterceptor/tagshelper.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	tracingComponentName = "yarpc-go"
+)
+
+// Static tracing tags to be used across spans
+var commonTracingTags = opentracing.Tags{
+	"go.version": runtime.Version(),
+	"component":  tracingComponentName,
+}
+
+// CommonTracingTags is the exported variable containing static tracing tags.
```
Comment (ChenX1993) [inline internal/tracinginterceptor/tagshelper.go:38]: It seems we don't need to make it exportable?

File: transport/tchannel/tracing.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+// PropagationCarrier is an interface to combine both reader and writer interface
+type PropagationCarrier interface {
+	opentracing.TextMapReader
+	opentracing.TextMapWriter
+}
+
+// GetPropagationCarrier get the propagation carrier depends on the transport.
+// The carrier is used for accessing the transport headers.
+// For TChannel, a special carrier is used. For details, see comments of HeadersCarrier
+func GetPropagationCarrier(headers map[string]string, transport string) PropagationCarrier {
```
Comment (ChenX1993) [inline transport/tchannel/tracing.go:43]: `GetPropagationCarrier` can also be put under `api/transport/propagation.go`, we just need tchannel's carrier to be here.

File: internal/tracinginterceptor/interceptor.go (hunk 1 above)
Comment (ChenX1993) [reply internal/tracinginterceptor/interceptor.go:137]: It doesn't seem to be resolved.

File: internal/interceptor/inbound.go [hunk 6]
```diff
@@ -24,7 +24,9 @@
 package interceptor
 
 import (
+	"context"
```
Comment (ChenX1993) [inline internal/interceptor/inbound.go:27]: Are those new changes under `internal/interceptor/...` really needed? I don't see any usage in other places in the PR.

File: internal/tracinginterceptor/interceptor_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChenX1993) [reply internal/tracinginterceptor/interceptor_test.go:33]:
```
out := yarpctest.NewFakeTransport().NewOutbound(nil,
		yarpctest.OutboundCallOverride(
			func(context.Context, *transport.Request) (*transport.Response, error) {
                             // you can do override here
                          },
		),
		yarpctest.OutboundCallStreamOverride(
			func(context.Context, *transport.StreamRequest) (*transport.ClientStream, error) {
                             // you can do override here
                          },
		),
		yarpctest.OutboundCallOnewayOverride(
			func(context.Context, *transport.Request) (transport.Ack, error) {
                             // you can do override here                          
                           },
		),
	)

tracingInterceptor.Call(ctx, req, out)
```

File: internal/tracinginterceptor/interceptor_test.go [hunk 7]
```diff
@@ -19,3 +19,151 @@
...
+	return nil
+}
+
+// Setup mock tracer
+func setupMockTracer() *mocktracer.MockTracer {
+	return mocktracer.New()
+}
+
+// TestUnaryInboundHandle tests the Handle method for Unary Inbound
+func TestUnaryInboundHandle(t *testing.T) {
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor_test.go:79]: Can we use the [test table](https://github.com/uber-go/guide/blob/master/style.md#test-tables) patter to test it more comprehensively like testing against successful case, failed case 1 (error is not nil), failed case 2(error is nil but with application error)? 

Same for outbound.

File: internal/tracinginterceptor/interceptor.go (hunk 1 above)
Comment (biosvs) [reply internal/tracinginterceptor/interceptor.go:137]: So, I see writer wrapper was copied here.

1. Does it mean, we'll have two wrappers now for the same request? So, one wrapper calls another wrapper that calls real handler?
2. Is there a plan to remove existing one? I guess, there are no?
3. As a side note: response size, for example, is not even in use, why was it copied here?

So, I guess we need to have one writer wrapper, it should be one layer of this header (which should be reused by obs and tracing).

File: internal/tracinginterceptor/interceptor.go [hunk 8]
```diff
@@ -39,32 +45,122 @@ var (
...
+		Tracer:        m.tracer,
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     CommonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, tchannel.GetPropagationCarrier(tracingHeaders, m.transport)); err != nil {
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:152]: Why GetPropagationCarrier is called from tchannel package?

File: internal/tracinginterceptor/interceptor.go [hunk 9]
```diff
@@ -39,32 +45,122 @@ var (
...
+		ext.Error.Set(span, true)
+		span.LogFields(log.String("event", "error"), log.String("message", err.Error()))
+		return nil, err
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:155]: I understand that it's existing logic, but should we actually return error for the call because tracing format is wrong?

File: internal/tracinginterceptor/interceptor.go [hunk 10]
```diff
@@ -39,32 +45,122 @@ var (
...
+	}
+
+	for k, v := range tracingHeaders {
+		req.Headers = req.Headers.With(k, v)
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:159]: Empty tracingHeaders was initialized at the line 151. What's happening here?

File: internal/tracinginterceptor/interceptor.go [hunk 11]
```diff
@@ -90,3 +186,48 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+	span.SetTag("error.type", "unknown_internal_yarpc")
+	return err
+}
+
+func updateSpanWithOutboundError(span opentracing.Span, res *transport.Response, err error) error {
+	isApplicationError := false
+	if res != nil {
+		isApplicationError = res.ApplicationError
+	}
+	if err == nil && !isApplicationError {
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:213]: Let's just use res.ApplicationError, why do we need isApplicationError?

File: internal/tracinginterceptor/interceptor.go [hunk 12]
```diff
@@ -90,3 +186,48 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+		return err
+	}
+
+	ext.Error.Set(span, true)
+	if yarpcerrors.IsStatus(err) {
+		status := yarpcerrors.FromError(err)
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:219]: IsStatus calls FromError inside, so we're doing casting twice.

Let's do it once:
```
if status, ok := FromError; ok {
}
```

File: internal/tracinginterceptor/interceptor.go [hunk 13]
```diff
@@ -90,3 +186,48 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+		span.SetTag("rpc.yarpc.status_code", errCode.String())
+		span.SetTag("error.type", errCode.String())
+		return err
+	}
+
+	span.SetTag("error.type", "unknown_internal_yarpc")
+	return err
+}
+
+func updateSpanWithOutboundError(span opentracing.Span, res *transport.Response, err error) error {
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:208]: Is it indeed for "outbound"? We probably may call it on inbound side, but when we already have response?

File: internal/tracinginterceptor/interceptor.go [hunk 14]
```diff
@@ -90,3 +186,48 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
 }
+
+func updateSpanWithError(span opentracing.Span, err error) error {
+	if err == nil {
+		return err
+	}
+
+	ext.Error.Set(span, true)
+	if yarpcerrors.IsStatus(err) {
+		status := yarpcerrors.FromError(err)
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:197]: Same comment about double casting as in updateSpanWithOutboundError

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/tracinginterceptor/interceptor.go (hunk 10 above)
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:159]: Updated with only adding tracing headers to req.Headers if the injection was successful

File: internal/tracinginterceptor/interceptor.go (hunk 9 above)
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:155]: This is a good call, thanks. I updated the code so that the error from m.tracer.Inject is logged and the tracing span is marked as an error, but the RPC call itself will still proceeds.

File: internal/tracinginterceptor/interceptor.go (hunk 13 above)
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:208]: Thanks, renamed to updateSpanWithResponseError

File: transport/tchannel/tracing/tracing.go [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+	"sync"
+)
+
+const (
+	tchannelTracingKeyPrefix      = "$tracing$"
+	tchannelTracingKeyMappingSize = 100
+)
+
+// PropagationCarrier is an interface to combine both reader and writer interface
+type PropagationCarrier interface {
```
Comment (ChenX1993) [inline transport/tchannel/tracing/tracing.go:35]: it seems there is not much value to define a new interface here.
We can just use the original two interfaces for interface compliance check:
```
var _ opentracing.TextMapReader = HeadersCarrier{}
var _ opentracing.TextMapWriter = HeadersCarrier{}
```

File: internal/tracinginterceptor/interceptor.go [hunk 16]
```diff
@@ -90,3 +148,63 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+	}
+
+	span.SetTag("error.type", "unknown_internal_yarpc")
+	return err
+}
+
+// GetPropagationFormat returns the opentracing propagation depends on transport.
+// For TChannel, the format is opentracing.TextMap
+// For HTTP and gRPC, the format is opentracing.HTTPHeaders
+func GetPropagationFormat(transport string) opentracing.BuiltinFormat {
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:195]: Nit: since we move those propagation related funcs to  the same package, there is no need to make them exported.

File: internal/tracinginterceptor/interceptor.go [hunk 17]
```diff
@@ -90,3 +148,63 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+func updateSpanWithResponseError(span opentracing.Span, res *transport.Response, err error) error {
+	if err == nil && (res == nil || !res.ApplicationError) {
+		return err
+	}
+
+	ext.Error.Set(span, true)
+	status := yarpcerrors.FromError(err)
+	if status != nil {
+		errCode := status.Code()
+		span.SetTag("rpc.yarpc.status_code", errCode.String())
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:178]: Don't we want to use the integer value for `rpc.yarpc.status_code` tag?

File: internal/tracinginterceptor/interceptor.go [hunk 18]
```diff
@@ -90,3 +148,63 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+	return err
+}
+
+func updateSpanWithResponseError(span opentracing.Span, res *transport.Response, err error) error {
+	if err == nil && (res == nil || !res.ApplicationError) {
+		return err
+	}
+
+	ext.Error.Set(span, true)
+	status := yarpcerrors.FromError(err)
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:175]: nit: more concise below since status is not used anywhere else
```
if status := yarpcerrors.FromError(err); status != nil {
}
```

File: internal/tracinginterceptor/interceptor.go [hunk 19]
```diff
@@ -39,32 +44,85 @@ var (
...
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         commonTracingTags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	err := h.Handle(ctx, req, resw)
+	if appErrSetter, ok := resw.(interface{ IsApplicationError() bool }); ok {
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:92]: The writer wrapper is probably still the right way to go with, the reasons are:
1. We also want to get the `applicationErrorMeta`
2. For the case where yarpc dispatcher is not used (hence obs middleware is not used), the original writer doesn't have `IsApplicationError()` method at all.

File: internal/tracinginterceptor/interceptor.go [hunk 20]
```diff
@@ -90,3 +148,63 @@ func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamReque
...
+		span.SetTag("rpc.yarpc.status_code", errCode.String())
+		span.SetTag("error.type", errCode.String())
+		return err
+	}
+
+	if res != nil && res.ApplicationError {
+		span.SetTag("error.type", "application_error")
+		return err
+	}
+
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:187]: Could we also check the `ApplicationErrorMeta` as well to add more error details?

File: internal/tracinginterceptor/interceptor.go (hunk 1 above)
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:99]: For now, we plan to implement two wrappers to minimize the workload. In the future, we aim to consolidate them into one.

File: internal/tracinginterceptor/interceptor.go (hunk 20 above)
Comment (biosvs) [reply internal/tracinginterceptor/interceptor.go:187]: Just as a reminder, we may use name and code, but should not add details.

File: internal/tracinginterceptor/interceptor.go (hunk 19 above)
Comment (biosvs) [reply internal/tracinginterceptor/interceptor.go:92]: Moreover, IsApplicationError() is implemented only for tChannel it seems. So tag value won't be set for other transports.

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/tracinginterceptor/interceptor.go [hunk 21]
```diff
@@ -22,10 +22,15 @@ package tracinginterceptor
 
 import (
 	"context"
+	"go.uber.org/yarpc/transport/tchannel/tracing"
```
[the remark this answers is not in the document]
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:25]: This hasn't been addressed in this commit, but I will address it eventually.

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/tracinginterceptor/interceptor.go [hunk 22]
```diff
@@ -39,54 +44,162 @@ var (
...
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     commonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, getPropagationCarrier(tracingHeaders, m.transport)); err != nil {
+		ext.Error.Set(span, true)
```
[the remark this answers is not in the document]
Comment (kexiongliu123) [reply internal/tracinginterceptor/interceptor.go:112]: Could you elaborate this a bit? The span should not have an error tag if Inject succeeds, and the error tag is only applied when Inject fails, which is already handled properly in the code

[the remark this answers is not in the document]
Comment (biosvs) [reply internal/tracinginterceptor/interceptor.go:116]: Yep, so, it's possible that Inject fail, but .Call succeed, right?

File: internal/tracinginterceptor/interceptor.go [hunk 23]
```diff
@@ -39,54 +45,184 @@ var (
...
+		span.SetTag("rpc.yarpc.status_code", int(errCode))
+		span.SetTag("error.type", errCode.String())
+		return err
+	}
+	if isApplicationError {
+		span.SetTag("error.type", "application_error")
+
+		if appErrorMeta != nil {
+			if appErrorMeta.Code != nil {
+				span.SetTag("application_error_code", int(*appErrorMeta.Code))
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:204]: Is it your plan to have integer code here? It would be unreadable from the UI, right?

File: transport/tchannel/handler.go [hunk 24]
```diff
@@ -49,12 +49,9 @@ type inboundCall interface {
 	ShardKey() string
 	RoutingKey() string
 	RoutingDelegate() string
-
```
Comment (biosvs) [inline transport/tchannel/handler.go:52]: Why all this blank lines are removed? Let's leave them, please. (Throughout the whole PR)

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/tracinginterceptor/interceptor.go [hunk 25]
```diff
@@ -39,54 +45,184 @@ var (
...
 func (m *Interceptor) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, h transport.UnaryHandler) error {
-	// TODO: implement
-	panic("implement me")
+	extendedWriter, ok := resw.(transport.ExtendedResponseWriter)
+	if !ok {
+		m.log.Debug("ResponseWriter does not implement ExtendedResponseWriter, proceeding without additional tracing")
+		return h.Handle(ctx, req, resw)
+	}
+
+	parentSpanCtx, _ := m.tracer.Extract(m.propagationFormat, getPropagationCarrier(req.Headers.Items(), req.Transport))
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:89]: Correct me if I am wrong - for outbound, `req.Transport` is set by the [firstoutboundmiddleware](https://github.com/yarpc/yarpc-go/blob/dev/internal/firstoutboundmiddleware/middleware.go#L74), so for the cases without yarpc dispatcher, the `req.Transport` will be empty.

Why not use `m.transport`?

File: internal/tracinginterceptor/interceptor.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (ChenX1993) [reply internal/tracinginterceptor/interceptor.go:116]: This is excellent point. We do want to use span error tag to indicate the success/failure statue of the real request. For the tracing own related failures, we also want to pop them up in some way.

Let's just remove the span error tag here but keep the span log. Internally, we have metrics for injection and extraction failure observability .

File: internal/tracinginterceptor/interceptor.go [hunk 26]
```diff
@@ -39,54 +45,184 @@ var (
...
+		StartTime:     time.Now(),
+		ExtraTags:     commonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, getPropagationCarrier(tracingHeaders, m.transport)); err != nil {
+		ext.Error.Set(span, true)
+		span.LogFields(log.String("event", "error"), log.String("message", err.Error()))
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:117]: Put `log.String("event", "error")` in `tagshelper.go`?  `var logFieldEventError`

File: internal/tracinginterceptor/interceptor.go (hunk 23 above)
Comment (ChenX1993) [reply internal/tracinginterceptor/interceptor.go:204]: For standard yarpc/http/gRPC/tchannel status code, we want to use int, and we plan to centrally convert them to human readable error code during ingestion time or query time.

For this, IIUC, appErrorMeta.Code is actually a yarpc code, so can we use the same name tag key `rpc.yarpc.status_code` if we want to keep it integer.

File: internal/tracinginterceptor/interceptor.go [hunk 27]
```diff
@@ -39,54 +45,184 @@ var (
...
+	}
+	if isApplicationError {
+		span.SetTag("error.type", "application_error")
+
+		if appErrorMeta != nil {
+			if appErrorMeta.Code != nil {
+				span.SetTag("application_error_code", int(*appErrorMeta.Code))
+			}
+			if appErrorMeta.Name != "" {
+				span.SetTag("application_error_name", appErrorMeta.Name)
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:207]: `error.name`, we already have `error.type=application_error` to indicate this is an application error, not need to include that in the tag key.

File: internal/tracinginterceptor/interceptor.go [hunk 28]
```diff
@@ -39,54 +45,184 @@ var (
...
+	err error,
+) error {
+	if err == nil && !isApplicationError {
+		return nil
+	}
+	ext.Error.Set(span, true)
+	if status := yarpcerrors.FromError(err); status != nil {
+		errCode := status.Code()
+		span.SetTag("rpc.yarpc.status_code", int(errCode))
+		span.SetTag("error.type", errCode.String())
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:196]: I know I put `error.type` in the ERD, but `error.code` seems to be more suitable here.

File: internal/tracinginterceptor/interceptor.go [hunk 29]
```diff
@@ -39,54 +45,184 @@ var (
...
+	appErrorMeta *transport.ApplicationErrorMeta,
+	err error,
+) error {
+	if err == nil && !isApplicationError {
+		return nil
+	}
+	ext.Error.Set(span, true)
+	if status := yarpcerrors.FromError(err); status != nil {
+		errCode := status.Code()
+		span.SetTag("rpc.yarpc.status_code", int(errCode))
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:195]: Can we define those strings as constants in `tagshelper.go`?

File: internal/tracinginterceptor/interceptor.go [hunk 30]
```diff
@@ -39,54 +45,184 @@ var (
...
+		Tracer:        m.tracer,
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     commonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, getPropagationCarrier(tracingHeaders, m.transport)); err != nil {
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:115]: Let's maybe add a code comment to explain why here we use m.transport, while we use req.Transport for inbound.

BTW, a small nit - for the receiver name, can we change to `i` instead of `m`, we use `m` because we originally called it tracing middleware, but it is interceptor now.

File: internal/tracinginterceptor/interceptor.go [hunk 31]
```diff
@@ -39,54 +45,189 @@ var (
...
+	}
+	m := &Interceptor{
+		tracer:            p.Tracer,
+		transport:         p.Transport,
+		propagationFormat: getPropagationFormat(p.Transport),
+		log:               p.Logger,
+	}
+	if m.tracer == nil {
+		m.tracer = opentracing.GlobalTracer()
+	}
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor.go:80]: 1. Let's change "m" name to "i", as it "interceptor now"
2. Let's put `if i.logger == nil` here too. It looks weird now: tracer is checked for interceptor, when logger is for params.

File: internal/tracinginterceptor/interceptor_test.go [hunk 32]
```diff
@@ -19,3 +19,519 @@
 // THE SOFTWARE.
 
 package tracinginterceptor
+
+import (
+	"context"
+	"fmt"
+	"github.com/opentracing/opentracing-go"
```
Comment (biosvs) [inline internal/tracinginterceptor/interceptor_test.go:26]: Have you rebased on last dev branch?

**Review by biosvs — CHANGES_REQUESTED** [review]

File: internal/tracinginterceptor/tagshelper.go [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"runtime"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/log"
+)
+
+const (
+	applicationErrorTag       = "application_error"
```
Comment (ChenX1993) [inline internal/tracinginterceptor/tagshelper.go:31]: nit - inconsistent, `applicationErrorTag` is a tag value similar to `unknownInternalYarpcError`, while `unknownInternalYarpcError` doesn't have `Tag` suffix. Let's remove the suffix

File: internal/tracinginterceptor/interceptor.go [hunk 34]
```diff
@@ -39,54 +45,189 @@ var (
...
+			if appErrorMeta.Code != nil {
+				span.SetTag(rpcStatusCodeTag, int(*appErrorMeta.Code))
+			}
+			if appErrorMeta.Name != "" {
+				span.SetTag(errorNameTag, appErrorMeta.Name)
+			}
+		}
+		return err
+	}
+	span.SetTag(errorCodeTag, unknownInternalYarpcError)
```
Comment (ChenX1993) [inline internal/tracinginterceptor/interceptor.go:217]: We have two sort of error code span tags here: 1) `rpc.yarpc.status_code`, 2) `error.code`. 
And we have 3 conditions here: 1) L198 - yarpc code derivable from err, 2) L204 - application err, 3) L217 - none of above

But they are not consistent:
* condition 1. sets tag 1) and tag 2)
* condition 2. sets tag 1)
* condition 3. sets tag 2)

Let's just just use tag 1) - `rpc.yarpc.status_code` for all of them.  So for condition 3, can we use 
```
span.SetTag(rpcStatusCodeTag, int(yarpcerrors.CodeUnknown))
```
And for condition 1, let's remove `error.code` tag.


For this general `error.code` span tag, let's add it during our ingestion pipeline, and use similar format like muttley error code: https://engwiki.uberinternal.com/display/TE0OBSERVAPPS/RPC+Error+Codes

**Review by ChenX1993 — APPROVED** [review]
Good job!

**Review by biosvs — APPROVED** [review]
